Tabs stop responding to all events

pvollan opened the pull request:
#### 7d88d417a15768253e791a3775ffbd31c951942b
<pre>
Tabs stop responding to all events
<a href="https://bugs.webkit.org/show_bug.cgi?id=296129">https://bugs.webkit.org/show_bug.cgi?id=296129</a>
<a href="https://rdar.apple.com/154677256">rdar://154677256</a>

Reviewed by Chris Dumez.

On rare occasions, tabs can stop responding to input events, and we see logging like this:

ViewState 2025-06-30 09:06:37.865762 [pageProxyID=27951, webPageID=27952, PID=11151] WebPageProxy::viewIsBecomingInvisible:
ViewState 2025-06-30 09:06:37.865747 [pageProxyID=27310, webPageID=27311, PID=10263] WebPageProxy::viewIsBecomingInvisible:
...
ViewState 2025-06-30 09:06:35.405510 [pageProxyID=27951, webPageID=27952, PID=11151] WebPageProxy::viewIsBecomingVisible:

From this logging we can see that initially, the WKWebView with WebContent PID 11151 becomes visible. Then the user likely
activated another tab, and then both WKWebViews become invisible. This is unexpected, since we would expect that the other
WKWebView becomes visible.

Code inspection shows that this can happen if PageClient::isViewVisible returns false for both WKWebViews in
WebPageProxy::updateActivityState. Further code inspection shows that this is possible, since the implementation in
PageClientImpl::isViewVisible on Mac returns the visibility of the active view, which isn&apos;t necessarily the WKWebView. It
can also be a _WKThumbnailView. From a WebPageProxy perspective, we are only interested in the visibility of the WKWebView,
and we shouldn&apos;t consider other views in that case.

* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::isActiveViewVisible):
(WebKit::PageClientImpl::isViewVisible): Deleted.
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::isActiveViewVisible):
(WebKit::PageClientImpl::isViewVisible): Deleted.
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::isMainViewVisible):
(WebKit::PageClient::isViewVisibleOrOccluded):
(WebKit::PageClient::isVisuallyIdle):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateActivityState):
(WebKit::WebPageProxy::activityStateDidChange):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::isActiveViewVisible):
(WebKit::PageClientImpl::isViewVisible): Deleted.
(WebKit::PageClientImpl::isViewWindowActive):
(WebKit::PageClientImpl::isViewVisibleOrOccluded):
(WebKit::PageClientImpl::isVisuallyIdle):
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::isViewVisible):
(WebKit::PageClientImpl::isActiveViewVisible):
(WebKit::PageClientImpl::isMainViewVisible):
(WebKit::PageClientImpl::isVisuallyIdle):
(WebKit::PageClientImpl::showCorrectionPanel):
(WebKit::PageClientImpl::showDictationAlternativeUI):

Canonical link: <a href="https://commits.webkit.org/297612@main">https://commits.webkit.org/297612@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/88de2d7087349ec7f623ac02cec411e536ed7a71

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112296 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32027 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22505 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118374 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62650 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32680 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40591 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/85309 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/36001 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-layout-tests-in-stress-mode; layout-tests (failure); Uploaded test results; layout-tests (failure); Compiled WebKit (cancelled)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115243 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26077 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101033 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65739 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/25380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62220 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/104810 "Build is in progress. Recent messages:") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/95461 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19248 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121700 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39370 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29299 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/94120 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39751 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97275 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93944 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39182 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/16976 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/35401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18103 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39258 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44746 "Built successfully") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38893 "Built successfully") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42230 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40636 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c1f87dd9629c6dd37d4193d59261746608b2f640. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48191

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f4b838635b3655feb9a4dd8fccb30a6b6e40e6ec. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48191

pvollan commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -217,6 +213,22 @@ - (NSCursor *)_cursorRectCursor;
+
+bool PageClientImpl::isMainViewVisible()
+{
+    RetainPtr activeView = m_view.get();
```
Ah, yes, will fix.

Thanks for reviewing!

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/66cc6db7aa5184f00c002ac2823727c9757e059d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48191

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5457d1031ae8446a0d7a8da8341f2091d9edca19. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48191

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0e29d63ef53e8dac2a373f1e37c6e18c0da82b91. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48191

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fe1c0d3e494660e5f0ec7d65d335351b76109d46. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48191

pvollan commented:
Thanks for reviewing!

pvollan committed:
Tabs stop responding to all events
https://bugs.webkit.org/show_bug.cgi?id=296129
rdar://154677256

Reviewed by Chris Dumez.

On rare occasions, tabs can stop responding to input events, and we see logging like this:

ViewState 2025-06-30 09:06:37.865762 [pageProxyID=27951, webPageID=27952, PID=11151] WebPageProxy::viewIsBecomingInvisible:
ViewState 2025-06-30 09:06:37.865747 [pageProxyID=27310, webPageID=27311, PID=10263] WebPageProxy::viewIsBecomingInvisible:
...
ViewState 2025-06-30 09:06:35.405510 [pageProxyID=27951, webPageID=27952, PID=11151] WebPageProxy::viewIsBecomingVisible:

From this logging we can see that initially, the WKWebView with WebContent PID 11151 becomes visible. Then the user likely
activated another tab, and then both WKWebViews become invisible. This is unexpected, since we would expect that the other
WKWebView becomes visible.

Code inspection shows that this can happen if PageClient::isViewVisible returns false for both WKWebViews in
WebPageProxy::updateActivityState. Further code inspection shows that this is possible, since the implementation in
PageClientImpl::isViewVisible on Mac returns the visibility of the active view, which isn't necessarily the WKWebView. It
can also be a _WKThumbnailView. From a WebPageProxy perspective, we are only interested in the visibility of the WKWebView,
and we shouldn't consider other views in that case.

* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::isActiveViewVisible):
(WebKit::PageClientImpl::isViewVisible): Deleted.
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::isActiveViewVisible):
(WebKit::PageClientImpl::isViewVisible): Deleted.
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::isMainViewVisible):
(WebKit::PageClient::isViewVisibleOrOccluded):
(WebKit::PageClient::isVisuallyIdle):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateActivityState):
(WebKit::WebPageProxy::activityStateDidChange):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::isActiveViewVisible):
(WebKit::PageClientImpl::isViewVisible): Deleted.
(WebKit::PageClientImpl::isViewWindowActive):
(WebKit::PageClientImpl::isViewVisibleOrOccluded):
(WebKit::PageClientImpl::isVisuallyIdle):
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::isViewVisible):
(WebKit::PageClientImpl::isActiveViewVisible):
(WebKit::PageClientImpl::isMainViewVisible):
(WebKit::PageClientImpl::isVisuallyIdle):
(WebKit::PageClientImpl::showCorrectionPanel):
(WebKit::PageClientImpl::showDictationAlternativeUI):

Canonical link: https://commits.webkit.org/297612@main

rkirsling commented:
This [broke](https://build.webkit.org/#/builders/65/builds/54419) the PlayStation build, as was indicated by EWS back [here](https://github.com/WebKit/WebKit/pull/48191#issuecomment-3089011916).

rkirsling commented:
I really should revert this on principle, because I need to deter it from happening again, but the fact is that it will take the same amount of time to fix as to revert, so I'm just going to fix it.
